pkg/openfeature/hooks_test.go

```diff
@@ -0,0 +1,762 @@
+package openfeature
+
+import (
+	"errors"
+	"reflect"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
+)
+
+func TestRequirement_4_1_1(t *testing.T) {
+	hookCtx := new(HookContext)
+
+	metaValue := reflect.ValueOf(hookCtx).Elem()
+
+	for _, name := range []string{"flagKey", "flagType", "evaluationContext", "defaultValue"} {
+		field := metaValue.FieldByName(name)
+		if field == (reflect.Value{}) {
+			t.Errorf("field %s doesn't exist in the HookContext struct", name)
+		}
+	}
+}
+
+func TestRequirement_4_1_2(t *testing.T) {
+	hookCtx := HookContext{}
+
+	type requirements interface {
+		ClientMetadata() ClientMetadata
+		ProviderMetadata() Metadata
+	}
+
+	var hookI interface{} = hookCtx
+	if _, ok := hookI.(requirements); !ok {
+		t.Error("HookContext doesn't implement the 4.1.2 requirements interface")
+	}
+}
+
+func TestRequirement_4_1_3(t *testing.T) {
+	hookCtx := new(HookContext)
+
+	metaValue := reflect.ValueOf(hookCtx).Elem()
+
+	caser := cases.Title(language.English)
+
+	for _, name := range []string{"flagKey", "flagType", "defaultValue"} {
+		field := metaValue.FieldByName(name)
+		if field == (reflect.Value{}) {
+			t.Errorf("field %s doesn't exist in the HookContext struct", name)
+		}
+
+		if caser.String(name) == name {
+			t.Errorf("field %s is uppercased and therefore mutable", name)
+		}
+	}
+}
+
+// Requirement_4_1_4 is satisfied by the evaluation context being immutable within the HookContext struct itself
+// and by the Before signature returning an evaluation context that is used to mutate the HookContext directly.
+
+func TestRequirement_4_2_1(t *testing.T) {
+	hookHints := HookHints{}
+
+	tpe := reflect.TypeOf(hookHints.mapOfHints)
+	if tpe.Kind() != reflect.Map {
+		t.Fatalf("expected HookHints kind to be map, got %s", tpe.Kind())
+	}
+	if tpe.Key().Kind() != reflect.String {
+		t.Errorf("expected HookHints key to be string, got %s", tpe.Key().Kind())
+	}
+	if tpe.Elem().Kind() != reflect.Interface {
+		t.Errorf("expected HookHints element to be interface{}, got %s", tpe.Elem().Kind())
+	}
+}
+
+func TestRequirement_4_2_2_1(t *testing.T) {
+	hookHints := new(HookHints)
+
+	metaValue := reflect.ValueOf(hookHints).Elem()
+
+	fieldName := "mapOfHints"
+	field := metaValue.FieldByName(fieldName)
+	if field == (reflect.Value{}) {
+		t.Errorf("field %s doesn't exist in the HookHints struct", fieldName)
+	}
+
+	caser := cases.Title(language.English)
+	if caser.String(fieldName) == fieldName {
+		t.Errorf("field %s is uppercased and therefore mutable", fieldName)
+	}
+}
+
+func TestRequirement_4_2_2_2(t *testing.T) {
```

Comment:
This might be tedious, but we might want to create some wrapper functions for tests or something that allow us to include the overall idea of the spec point (some text) instead of just the number. I only think this is important because the actual requirement represented by the numbers _could_ shift so it will help us keep things sane.

Probably something for another PR.

Reply:
I like this as an idea, I guess we could add the specification requirement as a comment to the test. To avoid the issue created by numbers shifting we could tie the sdk to a version of the spec.

Reply:
Created [an issue](https://github.com/open-feature/golang-sdk/issues/27) for this.